### ISSUE TITLE
The python bindings docs seems outdated

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#137]
Neither

http://wiki.opencog.org/w/Python

nor

https://github.com/opencog/atomspace/tree/master/opencog/python

work for me. They are somewhat contradictory as well. Obviously something is outdated in the python bindings documentations.

**Comment by ngeiswei** [comment]
OK it seems

https://github.com/opencog/atomspace/blob/master/opencog/cython/README.md

is correct. However the other docs should still be fixed. In particular
```
Install the required dependencies:
\```
cd ~/opencog/opencog/python
sudo pip install -r requirements.txt --upgrade
\```
```
in https://github.com/opencog/atomspace/blob/master/opencog/python/README.md doesn't work at all. And I assume this file should point the reader to https://github.com/opencog/atomspace/blob/master/opencog/cython/README.md for further details.

**Comment by ngeiswei** [comment]
I can try to fix it but I'm not totally confident. I'll create a pull request and let the experts merge...

**Comment by ngeiswei** [comment]
I've fixed the wiki to reflect Cython's README.md

http://wiki.opencog.org/w/Python#Setting_up

as well as the building page

http://wiki.opencog.org/w/Building_OpenCog#Building

**Comment by ngeiswei** [comment]
I think it can be closed but since I'm not sure about the requirements.txt thing as described in #138 I'll hand it back to the experts.

**Comment by linas** [comment]
Oh, I see. the requirements are ctually pre-reqs for the python modules, not the cython code.

Many of these prereqs are probably obsolete or no longer needed; they surely depend on the particular python module that someone wants to use.

**Comment by inflector** [comment]
@linas - do you want me to look at this list to narrow it down to the minimum? There may not actually be any requirements for just the atomspace. I can't think of why there would be.

**Comment by linas** [comment]
Are there any outstanding issues? I skimmed the wiki page, and the README's above, and did not see anything obviously flakey.  If its OK with @ngeiswei then maybe this issue can be closed.

**Comment by inflector** [comment]
@ngeiswei Python is working and I've fixed and tested http://wiki.opencog.org/w/Python line by line. I also updated it to reflect the output that one will get by running the examples with the current codebase.

So it's ready to be retested by someone else and closed.

**Comment by ngeiswei** [comment]
@inflector everything was mostly correct (I've fixed some minor discrepancies) till

http://wiki.opencog.org/wikihome/index.php/Python#MindAgents_in_Python

I'm getting the error
```
import opencog.cogserver
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
? in <module>()
----> 1 import opencog.cogserver

ImportError: /usr/local/lib/opencog/libserver.so: undefined symbol: _ZNK7opencog17StandardException10getMessageEv
```

**Comment by inflector** [comment]
I didn't touch or test anything from the Mind Agents and below. I'm not sure I understand that part. 

The context for running in the CogServer seems to be missing an introduction about the CogServer or a link over to a section about the CogServer and getting to the Python prompt within the CogServer. I will dig into this today and figure out what might be missing.

But I don't know if this has changed in the last year and much of the work I did last year was to remove the need for a CogServer in most cases when programming in Python.

When imported from the python prompt, I'm not getting that error. Have you done a recent:

`make`
`sudo make install`

on `atomspace`, `opencog`, and `cogutils`?

It looks like a linking error due to an outdated shared library import.

I believe this section only applies to python running in the cogserver Python prompt you get to via the `py` command, and not Python from the Python prompt one gets via bash the `python` command. Were you testing it from the conserver python prompt? Or the python prompt from bash directly?

**Comment by inflector** [comment]
There appear to be problems with the Python prompt in the CogServer as well. You can't even enter multi-line definitions as it won't terminate after a blank line like it should. This worked last year, I remember fixing this once already. I'll dig into it.

In reading it again, it appears that the MindAgents example is confusing because it shows entering the code with a interactive shell prompt when that isn't what it explains afterward, the code needs to be in a `some_mind_agent.py` file which gets loaded via the loudpy or PYTHON_PRELOAD variable.

**Comment by ngeiswei** [comment]
Oh, yes indeed I've run that from the python interpreter as nothing was explained about using the cogserver.

If nobody is current using Python in the cogserver maybe there is no need to fix it, just add a note in the wiki that this is WIP.

**Comment by linas** [comment]
I run python in the cogserver all the time. Its utterly and completely central to attaching opencog to ROS.  In particular, the code here: https://github.com/opencog/ros-behavior-scripting makes very heavy use of python, loaded from the cogsrever shell.  This code is integrated with the Hanson Robotics stack, and it would be disasterous if this code got broken.

**Comment by linas** [comment]
python mind-agents could probably be deprecated.  My general impression is that the whole concept of mind-agents needs a major overhaul.  The HR stack and the ROS behaviors, etc. do not use mind-agents; they simply launch new threads, as needed.

**Comment by linas** [comment]
A lsightly more self-contained demo of the cogserver plus python is the Eva chatbot, here: https://github.com/opencog/opencog/tree/master/opencog/nlp/chatbot-eva  I am actively developing this code now, and I expect it to work.

**Comment by bgoertzel** [comment]
@linas I think we all agree that the current  MindAgents framework is destined for near-term replacement.... @pennachin  and @hedayat  have been working on a replacement system that manages different agents each running in their own thread.   I think this work is almost ready for use now.   But so far as I know it doesn't involve python yet...

**Comment by linas** [comment]
"almost ready" ?? I haven't seen even a peep of ths on github yet.

**Comment by hedayat** [comment]
Then you've forgot about this PR: https://github.com/opencog/opencog/pull/1992

Actually, I consider it 'ready'. Since it is probably not used by anybody except me till now, it might have bugs though. There is one convenience feature that I'll push soon but it is not related to the core functionality.
Check this out:
```
opencog> help agents-start
Usage: agents-start {<agent type>[,<dedicated>[,<thread_name>]]} [...]

Create new agent instances of the specified agent type(s), and start them.
If <dedicated> is 'yes', the agent will run in a separate thread with optional name <thread_name>. If <thread_name> already exists, the agent will be added to that thread instead of a new thread.

```

**Comment by linas** [comment]
I just reviewed that code. There seems to be something deeply, fundamentally flawed in the design of that code.  A thread is created for each agent, but then that thread is blocked on a condition variable, preventing more than one thread from running at a time.   Then it gets worse ... 

The one running thread then calls the run() method on ALL of the other agents. Or tries to. Of course, if any one of the agents assumes that it is running in a thread, then the run() method will never-ever return, and the loop will deadlock.  Because the condition variable is still held, all of the other agents and threads will deadlock.

So the only way to have forward progress is to design the agent's run() method so that it returns BEFORE the agent is actually done.   All agents must do this, else there is no forward progress: any one is sufficent to deadlock all of them. And then, after all of these agents are run, then maybe (hard to tell, I stopped reading the code at this point) the condition is released, and now, this all happens again, except, this time, in some other thread.  

You are using threads and then forcing everything to be single-threaded: the condition variable blocks all threads.   The mapping is also not one-thread==one agent but many-threads, randomly accessing all agents.  The design doesn't make any sense to me. It is mind-bogglingly weird.

**Comment by hedayat** [comment]
There is a misunderstanding here, and I've tried to clarified it in the PR comments. 
I found that I've forgot to document the classes themselves to describe the purpose of the classes, so that there is a clear high-level view of the design. I'll fix it soon. 

BTW, while agents can have run() methods who doesn't return, they are not supposed to do so: we expect them to return as they do in the old design. 

The mapping is one-worker-thread -> one or more agents. But each agent is assigned to exactly one worker thread. Each worker thread only accesses its own agents.

**Comment by linas** [comment]
The "old design" is exactly what I'm arguing against.  Its a bad re-implementation of cooperative multi-tasking, which itself was considered to be obsolete when OS/2 came out in 1987, and again when Windows NT came out in 1993, and again when MacOSX came out 2001. Linux did preemptive multi-tasking from the  day it was born -- why are we trying to re-implement something that was considered to be obsolete a few decades ago?  

I mean, our "old design" is not even clever; its sophomoric, simple-minded.  MacOS 9 at least had it's event-driven subsystem, as a way of hiding some of the evils of coop multi-tasking. Later, green-threads made good use of setjmp/longjmp to save stack contexts and emulate continuations in user-space.  Recall that even green threads have bit the dust; even that was not enough to "do it right".

Why, on God's green Earth, would we even vaguely want to have something like this?  What possible technical reason could there possibly be?  Is there any technical benefit to this, whatsoever?

https://en.wikipedia.org/wiki/Computer_multitasking

**Comment by linas** [comment]
this bug "The python bindings docs seems outdated" is half a year old, and the conversatin about it has veered off-course.  I am closing this; if this is still a problem, please  open a new bug report.
